lib/BuildSystem/BuildFile.cpp

```diff
@@ -873,6 +921,61 @@ class BuildFileImpl {
       return nullptr;
     }
 
+    auto outputNodesAndCommands = std::vector<std::pair<BuildNode*, ExternalCommand*>>();
+    auto allDirectoryInputNodes = std::vector<BuildNode*>();
+
+    for (auto it = commands.begin(); it != commands.end(); it++) {
+      if ((*it).getValue()->isExternalCommand()) {
+        ExternalCommand* externalCommand = static_cast<ExternalCommand*>((*it).getValue().get());
+        for (auto output: externalCommand->getOutputs()) {
+          if (!output->isVirtual()) {
+            outputNodesAndCommands.push_back(std::pair<BuildNode*, ExternalCommand*>(output, externalCommand));
+          }
+        }
+
+        for (auto input: externalCommand->getInputs()) {
+          if (input->isDirectory()) {
+            allDirectoryInputNodes.push_back(input);
+          }
+        }
+      }
+    }
+
+    std::sort(outputNodesAndCommands.begin(), outputNodesAndCommands.end(), [](const std::pair<BuildNode*, ExternalCommand*> pairA, const std::pair<BuildNode*, ExternalCommand*> pairB) -> bool {
+      return pairA.first->getName().str().length() < pairB.first->getName().str().length();
+    });
+
+    OwnershipAnalysis ownershipAnalysis;
+
+    for (auto pair: outputNodesAndCommands) {
+      if (pair.first->isDirectory()) {
+        if (ownershipAnalysis.isUnownedNode(pair.first)) {
+          ownershipAnalysis.setOwner(pair.first, pair.second);
+        } else {
+          error("Multiple commands produce conflicting output.");
```

Comment:
Ensure the error message provides more context (e.g. which commands are in conflict?)